enhancements/kubelet/kubelet-credentials.md

```diff
@@ -0,0 +1,187 @@
+   `configmap/kubelet-bootstrap-kubeconfig`.
+4. The installer signs the master kubelet bootstrap credentials for `system:serviceaccount:openshift-machine-config-operator:node-bootstrapper`
+   with group `system:serviceaccounts:openshift-machine-config-operator`.
+   Notice that this matches the service account that will eventually be used.
+   The master kubelet bootstrap credentials are signed for 10 years of validity.
+5. In the `cluster-kube-apiserver-operator`, combined the content from `configmap/kubelet-bootstrap-kubeconfig` with other
+   valid client certificate CA bundles to identify clients.
+6. The master MachineConfig includes the master kubelet bootstrap credentials.
+
+To invalidate the master node bootstrap credential created by the installer
+1. Delete `configmap/kubelet-bootstrap-kubeconfig` in `ns/openshift-config`.
+2. The `cluster-kube-apiserver-operator` removes the CA bundle from the trusted client certificate CA bundles.
+3. The master node bootstrap credential created by the installer is no longer valid.
+4. In the MCO, if the `configmap/kubelet-bootstrap-kubeconfig` is deleted, use the "normal" service account token
+   logic that provides bootstrap authentication credentials for workers.
+5. The MCO rollouts the new machine configs.
+
+Worker node credentials are easier and look like this.
+1. The MCO retrieves a serviceaccount token for `serviceaccount/node-bootstrapper` in `ns/openshift-machine-config-operator`.
+2. The MCO includes this node bootstrap credential (the service account token) in MachineConfigs.
+   This ensures the content is distributed to each node during ignition.
+
+To invalidate the worker node credentials
+1. Delete the service account token: `oc -n openshift-machine-config-operator delete secret/node-bootstrapper-token`.
+2. The machine-config-operator will create a new, empty `node-bootstrapper-token` and the kube-controller-manager will
+   fill in a token value.
+3. The machine-config-operator will read the new token value and update the MachineConfig to distribute it to the nodes.
+
+### How Kubelets Create Credentials
+
+1. If a current kubelet client certificate exists and is valid, use it and skip the rest.
+2. Use the bootstrap kubelet credentials to create a client CertificateSigningRequest (CSR) on the kube-apiserver.
+   The username is in the form `system:node:<nodeName>` and groups contain `system:nodes`
+3. machine-approver (or cluster-admin) approves the CSR created by the kubelet's bootstrap credentials.
+4. Kubelet uses bootstrap kubelet credentials to retrieve the signed certificate from the client CSR.
+5. Given the signed client certificate and the local key (which never left the host), the kubelet communicate with the kube-apiserver
+
+### How machine-approver Decides to Approve Client Certificates
+
+1. If the CSR isn't a well-formed kubelet client certificate, don't approve.
+2. If the CSR is for a renewal, don't approve.
```

Comment:
We don't have any logic for 'renewals' specifically for client certs.  Step 4 more or less covers this.